### PR TITLE
fix: support custom servers using HTTP/2 in production

### PR DESCRIPTION
**Pull request by benmccann**
closes https://github.com/sveltejs/kit/issues/11365

### CODE REVIEW

**Review by Conduitry — COMMENTED** [review]
In #3572, we were able to check the major version of the incoming request and only operate on it then. Is that a possibility here, to avoid doing extra work cloning objects and looking for properties on every incoming request?

**Comment by benmccann** [comment]
done

I hadn't remembered that we had this before :laughing:  guess we should have kept it rather than merging https://github.com/sveltejs/kit/pull/7142. perhaps undici has gotten stricter than it used to be

File: .changeset/shy-needles-warn.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+'@sveltejs/kit': patch
+---
+
+fix: support custom servers using HTTP/2 in production
```
Comment (benmccann) [inline .changeset/shy-needles-warn.md:5]:
```suggestion
fix: support HTTP/2 in dev and production. Revert the changes from [#12907](https://github.com/sveltejs/kit/pull/12907) to downgrade HTTP/2 to TLS as now being unnecessary
```

File: packages/kit/src/exports/node/index.js [hunk 2]
```diff
@@ -106,11 +106,31 @@ function get_raw_body(req, body_size_limit) {
 // TODO 3.0 make the signature synchronous?
 // eslint-disable-next-line @typescript-eslint/require-await
 export async function getRequest({ request, base, bodySizeLimit }) {
+	let headers = /** @type {Record<string, string>} */ (request.headers);
+	if (request.httpVersionMajor >= 2) {
+		// the Request constructor rejects headers with ':' in the name
+		headers = Object.assign({}, headers);
+		if (headers[':method']) {
+			if (!headers.method) {
+				headers.method = headers[':method'];
```
[the remark this answers is not in the document]
Comment (benmccann) [reply packages/kit/src/exports/node/index.js]: oh yeah, you're right. thanks

for `:authority`, the spec actually says you have to do so when translating from http 2 to http 1.1. if you didn't, the host would be empty. i've updated it an added a comment

**Comment by eltigerchino** [comment]
This was something I tried in the previous PR but ran into some issues with the request URL being set incorrectly during https (https://github.com/sveltejs/kit/pull/12907#issuecomment-2443826591). Can you verify if this works with HTTPS turned on during dev / preview while navigating around a SvelteKit website?

**Comment by Rich-Harris** [comment]
preview: https://svelte-dev-git-preview-kit-12989-svelte.vercel.app/

_this is an automated message_

**Comment by benmccann** [comment]
It should work now, but please test and let me know if it doesn't

**Comment by benmccann** [comment]
If we merge https://github.com/sveltejs/kit/pull/12977 first it might help with testing

**Comment by eltigerchino** [comment]
LGTM. Managed to reproduce the issue I had earlier where the `host` was undefined. Had to replicate the HTTP2 `host` assignment logic from the `vite/dev/index.js` code to `vite/preview/index.js`.

https://github.com/sveltejs/kit/blob/7a17bd2712203ca8ec3a3343174a8e4185729bfa/packages/kit/src/exports/vite/dev/index.js#L440

https://github.com/sveltejs/kit/blob/7a17bd2712203ca8ec3a3343174a8e4185729bfa/packages/kit/src/exports/vite/preview/index.js#L188

**Review by eltigerchino — APPROVED** [review]

**Comment by benmccann** [comment]
Awesome. Thanks for the help!

**Pull request merged**